src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,468 @@
+    if let Ok(pk) = bitcoin::PublicKey::from_slice(&pk_bytes) {
+        let addr = bitcoin::Address::p2pkh(
+            &pk.to_public_key(),
+            bitcoin::Network::Bitcoin,
+        );
+        if *script_pubkey != addr.script_pubkey() || !witness.is_empty() {
+            return Err(Error::BadDescriptor)
+        }
+        Ok((Descriptor::Pkh(pk), stack))
+    } else{
+        Err(Error::BadDescriptor)
+    }
+}
+
+/// Helper for creating a p2sh descriptor based on script_pubkey and witness. Validates the p2sh
+/// hash based on redeem script, pops the redeem script from the script sig stack, translates other
+/// elements from scriptsig into Vec<StackElement>
+fn verify_p2sh<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &'txin bitcoin::Script,
+) -> Result<(Script, Vec<StackElement<'txin>>), Error>
+{
+    let (redeem_script, stack) = parse_scriptsig_top(script_sig)?;
+    let redeem_script = Script::from(redeem_script);
+    if redeem_script.to_p2sh() != *script_pubkey {
+        return Err(Error::BadDescriptor)
+    }
+    Ok((redeem_script, stack))
+}
+
+/// Figures out the the type of descriptor based on scriptpubkey, witness and scriptsig.
+/// Outputs a descriptor and witness_stack for all descriptors which can be directly fed into the
+/// interpreter. All script_sig and witness are translated into a single witness stack
+/// Vec<StackElement>
+/// 1) PK: creates a Pk descriptor and transfers the scriptsig to witness_stack<StackElement>
+/// 2) Pkh: Removes top element(pk) and validates pubkey hash, pushes rest of witness to
+/// witness_stack<StackElement> and outputs a Pkh descriptor
+/// 3) Wphk: translates witness to witness_stack<StackElement>, validates sig and pubkey hash
+/// and outputs a Wpkh descriptor
+/// 4) Wsh: pops witness script and checks wsh output hash, translates remaining witness elements
+/// to Vec<StackElement> and outputs a Wsh descriptor. Does not check miniscript inside the descriptor
+/// 5) Bare: translates script_sig to witness_stack<StackElement> and script_pubkey to miniscript
+/// 6) Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+/// witness_stack<StackElement> and redeem script to miniscript. Does not check the miniscript
+/// 7) ShWpkh: Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+/// witness_stack<StackElement> and validates Wpkh sig, pubkey.
+/// 8) ShWsh: Checks witness script hash, pops witness script and converts it to miniscript.
+/// translates the remaining witness to witness_stack<StackElement>
+pub fn witness_stack<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &'txin bitcoin::Script,
+    witness: &'txin [Vec<u8>],
+) -> Result<(Descriptor<PublicKey, hash160::Hash>, Vec<StackElement<'txin>>), Error>
+{
+    if script_pubkey.is_p2pk(){
+        verify_p2pk(script_pubkey, script_sig, witness)
+    }else if script_pubkey.is_p2pkh(){
```

Comment:
nit: `else if` should always be on the same line as the `}` and have a space before it. Similarly the `{` should have a space before it. (Same comment applies to all of the below.

Reply:
Actually never mind; I'll do a PR after this one which runs `cargo fmt` on everything and then CI-enforces that it passes in future.